pkg/controller/bastion/resourcesdefine.go

```diff
@@ -0,0 +1,161 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"encoding/base64"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-03-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-05-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	gardenerutils "github.com/gardener/gardener/pkg/utils"
+)
+
+func nicDefine(opt *Options, ip *network.PublicIPAddress, subnet *network.Subnet) *network.Interface {
+	return &network.Interface{
+		Name:     &opt.NicName,
+		Location: &opt.Location,
+		InterfacePropertiesFormat: &network.InterfacePropertiesFormat{
+			IPConfigurations: &[]network.InterfaceIPConfiguration{
+				{
+					Name: to.StringPtr("ipConfig1"),
+					InterfaceIPConfigurationPropertiesFormat: &network.InterfaceIPConfigurationPropertiesFormat{
+						Subnet: &network.Subnet{
+							ID: subnet.ID},
+						PrivateIPAllocationMethod: network.Dynamic,
+						PublicIPAddress:           ip,
+					},
+				},
+			},
+		},
+		Tags: opt.Tags,
+	}
+}
+
+func publicIPAddressDefine(opt *Options) *network.PublicIPAddress {
+	return &network.PublicIPAddress{
+		Name:     &opt.BastionPublicIPName,
+		Location: &opt.Location,
+		Sku: &network.PublicIPAddressSku{
+			Name: network.PublicIPAddressSkuNameStandard,
+		},
+		PublicIPAddressPropertiesFormat: &network.PublicIPAddressPropertiesFormat{
+			PublicIPAddressVersion:   network.IPv4,
+			PublicIPAllocationMethod: network.Static,
+		},
+		Tags: opt.Tags,
+	}
+}
+
+func computeInstanceDefine(opt *Options, bastion *extensionsv1alpha1.Bastion, adminPassword string) *compute.VirtualMachine {
+	return &compute.VirtualMachine{
+		Location: &opt.Location,
+		VirtualMachineProperties: &compute.VirtualMachineProperties{
+			HardwareProfile: &compute.HardwareProfile{
+				VMSize: compute.VirtualMachineSizeTypesStandardB1s,
+			},
+			StorageProfile: &compute.StorageProfile{
+				ImageReference: &compute.ImageReference{
+					Publisher: to.StringPtr("Canonical"),
+					Offer:     to.StringPtr("UbuntuServer"),
+					Sku:       to.StringPtr("18.04-LTS"),
+					Version:   to.StringPtr("latest"),
+				},
+				OsDisk: &compute.OSDisk{
+					CreateOption: compute.DiskCreateOptionTypesFromImage,
+					DiskSizeGB:   to.Int32Ptr(32),
+					Name:         &opt.DiskName,
+				},
+			},
+			OsProfile: &compute.OSProfile{
+				ComputerName:  &opt.BastionInstanceName,
+				AdminUsername: to.StringPtr("gardener"),
+				AdminPassword: &adminPassword,
```

Comment:
`Message=\"Required parameter 'adminPassword' is missing (null).\" Target=\"adminPassword\"`, as designed by Azure the AdminiPassword must be set. then we introduce the `password` function to generate random passwords.

```
type OSProfile struct {
    // ComputerName - Specifies the host OS name of the virtual machine. <br><br> This name cannot be updated after the VM is created. <br><br> **Max-length (Windows):** 15 characters <br><br> **Max-length (Linux):** 64 characters. <br><br> For naming conventions and restrictions see [Azure infrastructure services implementation guidelines](https://docs.microsoft.com/azure/azure-resource-manager/management/resource-name-rules).
    ComputerName *string `json:"computerName,omitempty"`
    // AdminUsername - Specifies the name of the administrator account. <br><br> This property cannot be updated after the VM is created. <br><br> **Windows-only restriction:** Cannot end in "." <br><br> **Disallowed values:** "administrator", "admin", "user", "user1", "test", "user2", "test1", "user3", "admin1", "1", "123", "a", "actuser", "adm", "admin2", "aspnet", "backup", "console", "david", "guest", "john", "owner", "root", "server", "sql", "support", "support_388945a0", "sys", "test2", "test3", "user4", "user5". <br><br> **Minimum-length (Linux):** 1  character <br><br> **Max-length (Linux):** 64 characters <br><br> **Max-length (Windows):** 20 characters.
    AdminUsername *string `json:"adminUsername,omitempty"`
    // AdminPassword - Specifies the password of the administrator account. <br><br> **Minimum-length (Windows):** 8 characters <br><br> **Minimum-length (Linux):** 6 characters <br><br> **Max-length (Windows):** 123 characters <br><br> **Max-length (Linux):** 72 characters <br><br> **Complexity requirements:** 3 out of 4 conditions below need to be fulfilled <br> Has lower characters <br>Has upper characters <br> Has a digit <br> Has a special character (Regex match [\W_]) <br><br> **Disallowed values:** "abc@123", "P@$$w0rd", "P@ssw0rd", "P@ssword123", "Pa$$word", "pass@word1", "Password!", "Password1", "Password22", "iloveyou!" <br><br> For resetting the password, see [How to reset the Remote Desktop service or its login password in a Windows VM](https://docs.microsoft.com/troubleshoot/azure/virtual-machines/reset-rdp) <br><br> For resetting root password, see [Manage users, SSH, and check or repair disks on Azure Linux VMs using the VMAccess Extension](https://docs.microsoft.com/troubleshoot/azure/virtual-machines/troubleshoot-ssh-connection)
    AdminPassword *string `json:"adminPassword,omitempty"`
    // CustomData - Specifies a base-64 encoded string of custom data. The base-64 encoded string is decoded to a binary array that is saved as a file on the Virtual Machine. The maximum length of the binary array is 65535 bytes. <br><br> **Note: Do not pass any secrets or passwords in customData property** <br><br> This property cannot be updated after the VM is created. <br><br> customData is passed to the VM to be saved as a file, for more information see [Custom Data on Azure VMs](https://azure.microsoft.com/blog/custom-data-and-cloud-init-on-windows-azure/) <br><br> For using cloud-init for your Linux VM, see [Using cloud-init to customize a Linux VM during creation](https://docs.microsoft.com/azure/virtual-machines/linux/using-cloud-init)
    CustomData *string `json:"customData,omitempty"`
    // WindowsConfiguration - Specifies Windows operating system settings on the virtual machine.
    WindowsConfiguration *WindowsConfiguration `json:"windowsConfiguration,omitempty"`
    // LinuxConfiguration - Specifies the Linux operating system settings on the virtual machine. <br><br>For a list of supported Linux distributions, see [Linux on Azure-Endorsed Distributions](https://docs.microsoft.com/azure/virtual-machines/linux/endorsed-distros).
    LinuxConfiguration *LinuxConfiguration `json:"linuxConfiguration,omitempty"`
    // Secrets - Specifies set of certificates that should be installed onto the virtual machine.
    Secrets *[]VaultSecretGroup `json:"secrets,omitempty"`
    // AllowExtensionOperations - Specifies whether extension operations should be allowed on the virtual machine. <br><br>This may only be set to False when no extensions are present on the virtual machine.
    AllowExtensionOperations *bool `json:"allowExtensionOperations,omitempty"`
    // RequireGuestProvisionSignal - Specifies whether the guest provision signal is required to infer provision success of the virtual machine.  **Note: This property is for private testing only, and all customers must not set the property to false.**
    RequireGuestProvisionSignal *bool `json:"requireGuestProvisionSignal,omitempty"`
}
```

Reply:
actually, there are two options here to set up ssh, also meets the minimum request from Azure as well, and both works. the `AdminUsername` Required parameter needs set in both solutions but doesn't matter what name is used, eventually, it is overwritten by userData.

option 1:   AdminUsername `gardener` or `gardener1` both works

```
OsProfile: &compute.OSProfile{
				ComputerName:  &opt.BastionInstanceName,
				AdminUsername: to.StringPtr("gardener"),
				AdminPassword: &adminPassword,  //must set and generate random passowrd 
}
UserData: to.StringPtr(base64.StdEncoding.EncodeToString(bastion.Spec.UserData)), 
```

option 2  AdminUsername `gardener` or `gardener1` both works
```
OsProfile: &compute.OSProfile{
				ComputerName:  &opt.BastionInstanceName,
				AdminUsername: to.StringPtr("gardener"),  
				LinuxConfiguration: &compute.LinuxConfiguration{
					DisablePasswordAuthentication: to.BoolPtr(true),
					SSH: &compute.SSHConfiguration{
						PublicKeys: &[]compute.SSHPublicKey{
							{
								Path:    to.StringPtr("/home/gardener/.ssh/authorized_keys"),  
								KeyData: to.StringPtr("ssh-rsa  xxxx"),  //fake ssh public key generated, the UserData will overwrite it later 
							},
						},
					},
				},
			},
UserData: to.StringPtr(base64.StdEncoding.EncodeToString(bastion.Spec.UserData)),
```

the current solution I used is option 1, let me know if we need switch to option 2

Reply:
forgot to mention last time. the SSH property Path must be set and not empty and KeyData must set and with valid ssh-rsa format 
change to option 2  ab34c249 with fake ssh ssh-rsa

Reply:
@tedteng I would not hardcode a ssh key, instead, please always generate a (throwaway) ssh public key to suffice the azure api

Reply:
fixed https://github.com/gardener/gardener-extension-provider-azure/pull/397/commits/0d5802edefe47fb22df7df78ec59b5c4d4fa9fb1